mwzeval/utils.py

```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
         json.dump(states, f, indent=2)
 
     return references, states
+
+
+def load_multiwoz24():
+    def is_filled(slot_value: str) -> bool:
+        """Whether a slot value is filled."""
+        slot_value = slot_value.lower()
+        return slot_value and slot_value != "not mentioned" and slot_value != "none"
+
+    def parse_state(turn: dict, prepend_book: bool = False) -> dict[dict[str, str]]:
+        """Get the slot values of a given turn.
+
+        This function is adapted from
+        google-research/schema_guided_dst/multiwoz/create_data_from_multiwoz.py
+
+        If a slot has multiple values (which are separated by '|'), only the first one is taken.
+        This is consistant with the approach taken for MultiWOZ 2.2 evaluation.
+
+        Args:
+            turn: Dictionary of a turn of the MultiWOZ 2.4 dataset
+            prepend_book: Whether to prepend the string 'book' to slot names for booking slots
+
+        Returns:
+            {$domain: {$slot_name: $value, ...}, ...}
+        """
+        dialog_states = defaultdict(dict)
+        for domain_name, values in turn['metadata'].items():
+            dialog_states_of_one_domain = {}
+
+            for k, v in values["book"].items():
+                # Note: "booked" is not really a state, just booking confirmation
+                if k == 'booked':
+                    continue
+                if isinstance(v, list):
+                    for item_dict in v:
+                        new_states = {
+                            (f"book{slot_name}" if prepend_book else slot_name): slot_val
```

Comment:
is this`bookpeople` or something like that in the corpus? or are we missing some dash etc?

Reply:
MultiWOZ 2.2 appends `book` to booking slots, e.g. in dialogue `PMUL2596`

MultiWOZ 2.2:
```
          {
            "actions": [],
            "service": "hotel",
            "slots": [],
            "state": {
              "active_intent": "book_hotel",
              "requested_slots": [],
              "slot_values": {
                "hotel-bookpeople": [
                  "4"
                ],
                "hotel-bookstay": [
                  "4"
                ],
                "hotel-name": [
                  "arbury lodge guesthouse"
                ]
              }
            }
          },
```

MultiWOZ 2.4:
```
                    "hotel": {
                        "book": {
                            "booked": [],
                            "stay": "4",
                            "day": "",
                            "people": "4"
                        },
                        "semi": {
                            "name": "arbury lodge guesthouse",
                            "area": "not mentioned",
                            "parking": "not mentioned",
                            "pricerange": "not mentioned",
                            "stars": "not mentioned",
                            "internet": "not mentioned",
                            "type": "not mentioned"
                        }
                    },
```

I don't think these prefixes are necessary for 2.4, and Google D3ST preprocessing script (https://github.com/google-research/task-oriented-dialogue/blob/main/state_tracking/utils/multiwoz_utils.py) actually removes the `book` prefix.

Reply:
crazy how complicated this became, it's ridiculous.